src/rest_vol.h

```diff
@@ -192,9 +192,198 @@
+                    HANDLE_RESPONSE(response_code, H5E_DATASET, H5E_WRITEERROR, FAIL);                       \
+                }                                                                                            \
+            }                                                                                                \
+        } /* end while (curl_multi_msg); */                                                                  \
+                                                                                                             \
+        /* TODO: Replace with an epoll-like structure of some kind, manually iterating this will probably    \
+         * be slow */                                                                                        \
+        struct timespec curr_time;                                                                           \
+        clock_gettime(CLOCK_MONOTONIC, &curr_time);                                                          \
+        size_t curr_time_ns = (size_t)curr_time.tv_sec * 1000 * 1000 * 1000 + (size_t)curr_time.tv_nsec;     \
+                                                                                                             \
+        for (size_t i = 0; i < count; i++) {                                                                 \
+            if (failed_handles_to_retry[i] &&                                                                \
+                ((curr_time_ns - time_of_fail[i]) >= current_backoff_duration[i])) {                         \
+                if (CURLM_OK != curl_multi_add_handle(curl_multi_handle, failed_handles_to_retry[i]))        \
+                    FUNC_GOTO_ERROR(H5E_DATASET, H5E_WRITEERROR, FAIL,                                       \
+                                    "failed to re-add denied cURL handle");                                  \
+                                                                                                             \
+                failed_handles_to_retry[i] = NULL;                                                           \
+            }                                                                                                \
+        }                                                                                                    \
+                                                                                                             \
+        /*                                                                                                   \
+        nanosleep(&delay, NULL);                                                                             \
+        */                                                                                                   \
+        num_prev_running = num_still_running;                                                                \
+    } while (num_still_running > 0);
```

Comment:
It's not a big deal at the moment, but I think this macro is at the point that it should just be a function. It's overly long for a macro and mixes in variables, memory management and more related to things that aren't passed in as a parameter and are assumed to be part of the function the macro is called in. For re-usability and readability, it should probably be refactored into a function at some point with necessary things being passed in. It's also usually better if the function allocating memory is the one freeing it, but that can be difficult in something like this where it's nice to free the memory as soon as we can.